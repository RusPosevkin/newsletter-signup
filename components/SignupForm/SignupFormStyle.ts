import styled, { keyframes } from "styled-components";

export const SignupFormContainer = styled.div`
  display: flex;
  align-items: center;
  justify-content: center;
  width: 100%;
  padding-top: 1rem;
`;

export const TermsLink = styled.a`
  color: rgb(0, 150, 136);
  text-decoration: underline;
`;

export const EmailInput = styled.input`
  width: 180px;
  font-size: 0.9rem;
  border-radius: 0.3rem;
  background-color: rgb(255, 255, 255);
  border: 1px solid rgb(0, 150, 136);
`;

export const CategorySelect = styled.select`
  width: 190px;
  font-size: 0.9rem;
  border-radius: 0.3rem;
  background-color: rgb(255, 255, 255);
  border: 1px solid rgb(0, 150, 136);
`;

export const SubmitButton = styled.button`
  width: 100%;
  font-size: 0.9rem;
  border-radius: 0.3rem;
  background-color: rgb(255, 255, 255);
  border: 1px solid rgb(0, 150, 136);
`;

export const FormSection = styled.section`
  display: block;
  margin-bottom: 1rem;
`;

export const ErrorMessage = styled.section`
  display: block;
  color: red;
`;

export const Label = styled.label<{ required?: boolean; }>`
  &:after {
    content: ${(props) => `"${props.required ? ' * ' : ''}"`};
    color: red;
  }
`;

const animation = keyframes`
 to {
    background-position: left;
  }
`
export const Skeleton = styled.div`
  display: block;
  cursor: progress;
  width: 300px;
  height: 1em;
  background: linear-gradient(90deg, #ddda 40%, #efefefaa, #ddda 60%) right / 300% 100%;
  animation: ${animation} 1s linear infinite;
  margin: 20px 0;
`
