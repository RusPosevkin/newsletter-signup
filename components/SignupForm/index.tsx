import SignupFormComponent from './SignupFormStructure';

export default SignupFormComponent;
