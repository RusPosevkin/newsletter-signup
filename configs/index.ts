import { DefaultTheme } from "styled-components";

export const theme: DefaultTheme = {
  colors: {
    primary: "#111",
    secondary: "#0070f3",
    alert: '#f1fdf1'
  },
};

// time in milliseconds for showing alert message
export const ALERT_DELAY_TIMEOUT = 5000;

export const ALERT_ANIMATION_DELAY = 500;
