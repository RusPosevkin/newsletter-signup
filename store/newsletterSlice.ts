import { createSlice } from "@reduxjs/toolkit";
import type { PayloadAction } from "@reduxjs/toolkit";

type NewsletterState = {
  isSubmitted: boolean;
  message: string;
};

export interface INewsletterState {
  newsletterState: NewsletterState
}

const initialState: INewsletterState = {
  newsletterState: {
    isSubmitted: false,
    message: ''
  }
};

export const newsletterSlice = createSlice({
  name: 'newsletter',
  initialState,
  reducers: {
    setNewsletterState: (state, action: PayloadAction<NewsletterState>) => {
      state.newsletterState = action.payload;
    },
  },
});

export const { setNewsletterState } = newsletterSlice.actions;
export const newsletterReducer = newsletterSlice.reducer;
