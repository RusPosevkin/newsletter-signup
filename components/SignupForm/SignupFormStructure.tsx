import React, { ChangeEvent, FC, FormEvent, Key, useRef, useState } from 'react'
import Link from 'next/link';
import { ErrorMessage, FormSection, Label, SignupFormContainer, TermsLink } from "./SignupFormStyle";
import { useAppDispatch } from '@/hooks/store';
import { setNewsletterState } from '@/store/newsletterSlice';
import { getAlertMessage, validateEmail, validateSignUpForm } from '@/services';

const SignupFormStructure: FC<{ termsUrl?: string }> = ({ termsUrl = '#' }) => {
  const [validationErrors, setValidationErrors] = useState<Array<String>>([]);
  const dispatch = useAppDispatch();
  const formRef = useRef<HTMLFormElement>(null);

  const handleSubmit = (e: FormEvent<HTMLFormElement>) => {
    e.preventDefault();
    setValidationErrors([]);

    const form = e.currentTarget;
    const formData = new FormData(form);
    const formJson = Object.fromEntries(formData.entries());
    console.log(formJson);
    const {
      isValid,
      messages
    } = validateSignUpForm(formJson);

    if (!isValid) {
      setValidationErrors(messages);
      return;
    }

    console.log('submit form');

    const message = getAlertMessage(formJson);

    dispatch(setNewsletterState({
      message,
      isSubmitted: true
    }));

    // reset form
    formRef.current && formRef.current.reset();
  }

  return (
    <SignupFormContainer>
      <form onSubmit={handleSubmit} ref={formRef}>
        <FormSection>
          <Label required htmlFor="email">
            Your email
          </Label>
          <input
            id="email"
            name="email"
            data-testid="input-email"
            placeholder='Enter your email address'
          />
        </FormSection>
        <FormSection>
          <Label>
            I'm interested in { }
            <select name="interestCategory" defaultValue="default" data-testid="select-category">
              <option value="default" disabled>choose a category</option>
              <option value="shoes">Shoes</option>
              <option value="heels">Heels</option>
              <option value="sandles">Sandles</option>
              <option value="handbags">Handbags</option>
            </select>
          </Label>
        </FormSection>
        <FormSection>
          <input type="checkbox" name="terms" id="terms" data-testid="checkbox-terms" />
          <Label required htmlFor="terms">
            I agree to the { }
            <Link href={termsUrl} passHref legacyBehavior>
              <TermsLink>terms and conditions</TermsLink>
            </Link>
          </Label>
        </FormSection>
        {
          validationErrors.length > 0 && (<ErrorMessage>
            {validationErrors.map((error, index) => (
              <p key={index}>
                {error}
              </p>
            ))}
          </ErrorMessage>)
        }
        <button type="submit" data-testid="button-submit">Subscribe</button>
      </form>
    </SignupFormContainer>
  );
}

export default SignupFormStructure;
