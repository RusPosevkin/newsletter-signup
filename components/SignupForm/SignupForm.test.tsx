import { fireEvent, render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import { ThemeProvider } from 'styled-components';
import { Provider } from 'react-redux';
import { theme } from '@/configs';
import { store } from '@/store/store';
import SignupForm from '.'

const withProviders = (children: React.JSX.Element) => (
  <ThemeProvider theme={theme}>
    <Provider store={store}>{children}</Provider>
  </ThemeProvider>
);

const filmsDataMock = {
  data: ['Film 1', 'Film 2'],
  loading: false
}

test('it should match the snapshot and render', async () => {
  const { container } = render(
    withProviders(
      <SignupForm films={filmsDataMock} />
    )
  );

  expect(container).toMatchSnapshot();

  const emailInput = screen.getByTestId('input-email');
  const categorySelect = screen.getByTestId('select-category');
  const termsCheckbox = screen.getByTestId('checkbox-terms');
  const submitButton = screen.getByTestId('button-submit');

  expect(container).toBeInTheDocument();
  expect(emailInput).toBeInTheDocument();
  expect(categorySelect).toBeInTheDocument();
  expect(termsCheckbox).toBeInTheDocument();
  expect(submitButton).toBeInTheDocument();
});

test('it show loading page when films are loading', async () => {
  const filmsDataLoadingMock = {
    ...filmsDataMock,
    loading: true
  }
  render(
    withProviders(
      <SignupForm films={filmsDataLoadingMock} />
    )
  );

  const emailInput = screen.queryByTestId('input-email');
  const categorySelect = screen.queryByTestId('select-category');
  const termsCheckbox = screen.queryByTestId('checkbox-terms');
  const submitButton = screen.queryByTestId('button-submit');

  const skeletonLoadingSection = screen.getByTestId('skeleton-loading');

  expect(emailInput).not.toBeInTheDocument();
  expect(categorySelect).not.toBeInTheDocument();
  expect(termsCheckbox).not.toBeInTheDocument();
  expect(submitButton).not.toBeInTheDocument();

  expect(skeletonLoadingSection).toBeInTheDocument();
});

test('it should validate empty email and not accepted terms and conditions', async () => {
  render(
    withProviders(
      <SignupForm films={filmsDataMock} />
    )
  );

  const submitButton = screen.getByTestId('button-submit');

  fireEvent.click(submitButton);

  expect(screen.getByText('Please enter your email address')).toBeInTheDocument();
  expect(screen.getByText('Please accept terms and conditions')).toBeInTheDocument();
});

test('it should validate incorrect email', async () => {
  render(
    withProviders(
      <SignupForm films={filmsDataMock} />
    )
  );

  const emailInput = screen.getByTestId('input-email');
  const termsCheckbox = screen.getByTestId('checkbox-terms');
  const submitButton = screen.getByTestId('button-submit');

  fireEvent.change(emailInput, { target: { value: 'incorrect' } });
  fireEvent.click(termsCheckbox);
  fireEvent.click(submitButton);

  expect(screen.getByText('Invalid email format')).toBeInTheDocument();
});

test('it should reset previous validation errors between validations', async () => {
  render(
    withProviders(
      <SignupForm films={filmsDataMock} />
    )
  );

  const submitButton = screen.getByTestId('button-submit');
  const emailInput = screen.getByTestId('input-email');
  const termsCheckbox = screen.getByTestId('checkbox-terms');

  fireEvent.click(submitButton);

  expect(screen.getByText('Please enter your email address')).toBeInTheDocument();
  expect(screen.getByText('Please accept terms and conditions')).toBeInTheDocument();

  fireEvent.change(emailInput, { target: { value: 'incorrect' } });
  fireEvent.click(termsCheckbox);

  fireEvent.click(submitButton);

  expect(screen.getByText('Invalid email format')).toBeInTheDocument();
});

test('it should reset form after successful validation and form submit', async () => {
  render(
    withProviders(
      <SignupForm films={filmsDataMock} />
    )
  );

  const submitButton = screen.getByTestId('button-submit');
  const emailInput = screen.getByTestId('input-email');
  const categorySelect = screen.getByTestId('select-category');
  const termsCheckbox = screen.getByTestId('checkbox-terms');

  fireEvent.change(emailInput, { target: { value: 'correct@email.com' } });
  fireEvent.change(categorySelect, { target: { value: 'heels' } });
  fireEvent.click(termsCheckbox);

  expect(emailInput).toHaveValue('correct@email.com');
  expect(categorySelect).toHaveValue('heels');
  expect(termsCheckbox).toBeChecked();

  fireEvent.click(submitButton);

  expect(emailInput).toHaveValue('');
  expect(categorySelect).toHaveValue('default');
  expect(termsCheckbox).not.toBeChecked();
});

test('it should add film options in select control', async () => {
  render(
    withProviders(
      <SignupForm films={filmsDataMock} />
    )
  );

  expect(screen.getByRole('option', { name: /Film 1/i })).toBeInTheDocument();
  expect(screen.getByRole('option', { name: /Film 2/i })).toBeInTheDocument();
});

