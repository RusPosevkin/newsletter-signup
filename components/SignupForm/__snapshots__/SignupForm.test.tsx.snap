// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it should match the snapshot and render 1`] = `
<div>
  <div
    class="sc-ae70ea4f-0 dvPLfn"
  >
    <form>
      <section
        class="sc-ae70ea4f-2 emiEpD"
      >
        <label
          class="sc-ae70ea4f-4 VmMEr"
          for="email"
          required=""
        >
          Your email
        </label>
        <input
          data-testid="input-email"
          id="email"
          name="email"
          placeholder="Enter your email address"
        />
      </section>
      <section
        class="sc-ae70ea4f-2 emiEpD"
      >
        <label
          class="sc-ae70ea4f-4 bfFRit"
        >
          I'm interested in 
          <select
            data-testid="select-category"
            name="interestCategory"
          >
            <option
              disabled=""
              selected=""
              value="default"
            >
              choose a category
            </option>
            <option
              value="shoes"
            >
              Shoes
            </option>
            <option
              value="heels"
            >
              Heels
            </option>
            <option
              value="sandles"
            >
              Sandles
            </option>
            <option
              value="handbags"
            >
              Handbags
            </option>
          </select>
        </label>
      </section>
      <section
        class="sc-ae70ea4f-2 emiEpD"
      >
        <input
          data-testid="checkbox-terms"
          id="terms"
          name="terms"
          type="checkbox"
        />
        <label
          class="sc-ae70ea4f-4 VmMEr"
          for="terms"
          required=""
        >
          I agree to the 
          <a
            class="sc-ae70ea4f-1 kJNCBL"
            href="#"
          >
            terms and conditions
          </a>
        </label>
      </section>
      <button
        data-testid="button-submit"
        type="submit"
      >
        Subscribe
      </button>
    </form>
  </div>
</div>
`;
